src/libraries/System.Private.CoreLib/src/System/Buffers/Utilities.cs

```diff
@@ -14,8 +14,7 @@ internal static class Utilities
         internal static int SelectBucketIndex(int bufferSize)
         {
             Debug.Assert(bufferSize >= 0);
-            uint bits = ((uint)bufferSize - 1) >> 4;
-            return 32 - BitOperations.LeadingZeroCount(bits);
+            return BitOperations.Log2Unsafe((uint)bufferSize - 1 | 0xF) - 3;
```

Comment:
I left the `Log2Unsafe` in a few places for further discussion.  This use is an example where the change from `LeadingZeroCount` to `Log2` is technically more expensive when `lzcnt` is supported by the processor.  The original sequence would have been `sub/rsh/lzcnt/neg/add`, where the new sequence would be `sub/or/or/lzcnt/xor/sub` with `Log2`.  The use of `Log2Unsafe` eliminates the redundant `or`, making them equal.

On processors not supporting `lzcnt` (or in R2R CoreLib), the sequence changes from `sub/rsh/bsr/xor/neg/add` to `sub/or/or/bsr/sub` with `Log2`, which is an always an improvement and can be further improved by eliminating the redundant `or` with `Log2Unsafe`.

I was unable to measure any difference in `ArrayPool<T>` Rent/Return with `Log2` vs `Log2Unsafe`,  but I didn't want to replace what was there with something more expensive.  IMO, using `Log2` here makes the intent of the code much more clear, so outside the performance considerations, I thought it was a good change.

Reply:
> IMO, using `Log2` here makes the intent of the code much more clear,

I do not think `Log2Unsafe` makes the intent of the code more clear in this case. I believe that the original code is much easier to understand.

Reply:
How so?  The bucket index is defined as log<sub>2</sub> of the buffer length, with the exception that the first 3 buckets (length 1-16) are combined, and the index is slid down by 3 to compensate.  The updated code says that exactly.